grafana/grafana
# RemoteCache: Add distributed in memory cache backend

**What is this feature?**
As a hackton project me and @Jguer worked on a new backend for remote cache. This is a distributed in memory cache using a consistent hash ring from dskit.

This is a useful alternative to the default one (database) when users don't want to set up additional alternatives such as redis or memchache. But it requires some configuration to be able to form the cluster so we can't make this the default value. Without proper configuration it would act as a in memory cache.

To test it out locally you can run:
```bash
make build-go
GF_DEFAULT_APP_MODE=development GF_SERVER_HTTP_ADDR=127.0.0.1 GF_GRPC_SERVER_ADDRESS="127.0.0.1:5000" GF_REMOTE_CACHE_TYPE=ring ./bin/linux-amd64/grafana server
```
This would start the initial node.

For all other nodes we need to omit `GF_DEFAULT_APP_MODE=development` because this binds a hardcoded address + port so they wont start. You would also need to set `GF_REMOTE_CACHE_RING_JOIN_MEMBERS` with at least one other member of the cluster and use another addres.

E.g. a second node can be started with
```bash
GF_SERVER_HTTP_ADDR=127.0.0.2 GF_GRPC_SERVER_ADDRESS="127.0.0.2:5000" GF_REMOTE_CACHE_TYPE=ring GF_REMOTE_CACHE_RING_JOIN_MEMBERS='127.0.0.1:4000' ./bin/linux-amd64/grafana server
```
Because the initial node is started in "dev" mode we expose useful debug endpoints for the ring so we can check the status of it: `127.0.0.1:3000/remote_cache/ring/kv` and `http://localhost:3000/remote_cache/ring/status`

**Which issue(s) does this PR fix?**:


Fixes #

**Special notes for your reviewer:**

Please check that:
